pymeasure/instruments/temptronic/temptronic_base.py

```diff
@@ -0,0 +1,841 @@
+                self.set_point_number = 0 # hot
+            else:
+                raise ValueError(f"Temperature {set_temp} is impossible to set!")
+
+        self.temp_set_point=set_temp 
+
+    def wait_for_settling(self, time_limit=300, log=False):
+        """sweep to a specified.
+
+        Parameters
+        ----------
+        set_temp : float
+            target temperature for DUT
+        
+        time_limit : float
+            set the maximum blocking time within thermostream has to settle.
+            Script execution is blocked until either thermostream has settled
+            or time_limte has been exceeded.
+        
+        log : boolean
+            show polling progress in command line
+            
+        Returns
+        -------
+        none
+        
+        """
+
+        time.sleep(1)
+        t = 0
+        while( not self.at_temperature()): #assert at temperature
+            time.sleep(1)
+            t += 1
+            if log == True:
+                print(f"temp_set= {self.temperature_setpoint:4.1f} deg,",
+                      f"temp = {self.temperature:4.1f} deg,",
+                      f"time elapsed = {t:4.0f} s,",
+                      f"status = {self.temperature_event_status}")
+                
+            if t > time_limit:
+                print('no settling achieved')
+                break
+
+    def shutdown(self):
+        """Brings the instrument to a safe and stable state"""
+        self.isShutdown = True
+        log.info("Shutting down %s" % self.name)
+        self.enable_air_flow = 0
+        self.remote_mode = False
+        self.head = 'up'
+
+    def start(self):
+        self.remote_mode = True
+        self.enable_air_flow = 1 # enable thermostream
+
+    def cycling_stopped(self):
+        """check if cycling has stopped.      
+        """
+        return self.temperature_condition_status_code & TemperatureStatusCode.CYCLING_STOPPED
+
+    def end_of_all_cycles(self):
+        """ check if cycling has stopped.
+        """
+        return self.temperature_condition_status_code & TemperatureStatusCode.END_OF_ALL_CYCLES
+
+    def end_of_one_cycle(self):
+        """ check if end of (at least) one cycle is done.
+        """
+        return self.temperature_condition_status_code & TemperatureStatusCode.END_OF_ONE_CYCLE
+
+    def end_of_test(self):
+        """ check if end of test is reached.
+        """
+        return self.temperature_condition_status_code & TemperatureStatusCode.END_OF_TEST
+
+    def not_at_temperature(self):
+        """ check if not (yet) at temperature.
+        """
+        return self.temperature_condition_status_code & TemperatureStatusCode.NOT_AT_TEMPERATURE
+
+    def at_temperature(self):
+        """ check if at temperature.
+        """
+        return self.temperature_condition_status_code & TemperatureStatusCode.AT_TEMPERATURE
+    
+    def __enter__(self):
+        return self
+
+    def __exit__(self):
+        self.shutdown()
```

Comment:
Very nice idea! @bilderbuchi: would it be a good idea to add this to the standard implementation of the `Instruments` class?

Reply:
Yes, in principle it's nice to enable using instruments with a `with` statement. 
However, I think many users don't use instruments like that, but instead create an instrument, run the whole experiment/scripted measurement/whatever, and destroy the instrument at the end. In my experience this would mean nest the whole code into a `with` statement for every used instrument, which makes the code structure not very nice. 

Additionally, for most instruments the `shutdown` method is basically a no-op (and what a "safe" shutdown means might be specific to most setups/experiments), so we would not gain much of the safety enabled by with statements.

Reply:
In this particular case I use the with statement to safeguard the measurement setup. During the unsupervised tests I set the thermostream to 225°C for a few seconds. If my program crashes during that time and nobody is around it will melt down my PCB and might start to burn setup. For me it's not jut nice to have but a safety feature for this piece of equipment.

Reply:
Sure, it's well suited for that. You could also achieve the same safety with a `try` - `finally`, without needing this specific class feature, though.

Reply:
Let's keep this here for now, if this pattern gets more usage we can always roll it out to the generic instrument class.

Reply:
This has been added to `Instrument` by now.